hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala

```diff
@@ -114,10 +112,33 @@ class PathOutputCommitProtocol(
         // failures. Warn
         logTrace(s"Committer $committer may not be tolerant of task commit failures")
       }
+    } else {
+      // if required other committers need to be checked for dynamic partition
+      // compatibility through a StreamCapabilities probe.
+      if (dynamicPartitionOverwrite) {
+        if (supportsDynamicPartitions) {
+          logDebug(
+            s"Committer $committer has declared compatibility with dynamic partition overwrite")
+        } else {
+          throw new IOException(PathOutputCommitProtocol.UNSUPPORTED + ": " + committer)
+        }
+      }
     }
     committer
   }
 
+
+  /**
+   * Does the instantiated committer support dynamic partitions?
+   * @return true if the committer declares itself compatible.
+   */
+  private def supportsDynamicPartitions = {
+    committer.isInstanceOf[FileOutputCommitter] ||
```

Comment:
Do we change the old behaviour because of  the `committer.isInstanceOf[FileOutputCommitter]` part of the condition? As before this PR we were simply throwing an exception.
If this is a change then the `Does this PR introduce any user-facing change?` section of the PR description must be updated too.

Reply:
up until now, if you used file output committer through the factory, it would reject the use of dynamic partitions. now it lets the caller without needing to implement the path capability.

this does ensure that you can use that committer through the factory mechanism even on hadoop versions <= 3.3.4. I didn't add the path capabilities probe to the original FileOutputCommitter because I'm scared of going near that file: old, complex, critical code

+ will update the user facing details